File: c++/graphLinkedList.cpp
#include <iostream>
#include <vector>

using namespace std;

enum GraphType
{
    DIRECTED,  // Graf Terarah
    UNDIRECTED // Graf Tidak Terarah
};

class GraphLinkedList
{
public:
    int vertices;
    vector<vector<int>> list;
    GraphType graphType;

    GraphLinkedList(int vertices, GraphType graphType)
    {
        this->vertices = vertices;
        this->graphType = graphType;
        for (int i = 0; i < vertices; i++)
            this->list.push_back(vector<int>());
    }

    void addEdge(int i, int j)
    {
        this->list[i].push_back(j);
        if (this->graphType == GraphType::UNDIRECTED)
            this->list[j].push_back(i);
    }

    void printGraph()
    {
        for (int i = 0; i < this->vertices; i++)
        {
            cout << i;
            for (int j = 0; j < this->list[i].size(); j++)
            {
                cout << " -> ";
                cout << this->list[i][j];
            }
            cout << endl;
        }
    }
};

int main(int argc, char const *argv[])
{
    GraphLinkedList graph(4, GraphType::UNDIRECTED);
    graph.addEdge(1, 2);
    graph.addEdge(1, 3);
    graph.printGraph();

    return 0;
}

File: java/GraphLinkedList.java
import java.util.ArrayList;
import java.util.LinkedList;

public class GraphLinkedList {
    enum GraphType {
        DIRECTED, // Graf terarah
        UNDIRECTED // Graf tidak terarah
    }

    int vertices;
    ArrayList<LinkedList<Integer>> list;
    GraphType graphType;

    GraphLinkedList(int vertices, GraphType graphType) {
        this.vertices = vertices;
        this.graphType = graphType;
        this.list = new ArrayList<>(vertices);
        for (int i = 0; i < vertices; i++)
            list.add(new LinkedList<Integer>());
    }

    void addEdge(int i, int j) {
        list.get(i).add(j);
        if (this.graphType == GraphType.UNDIRECTED)
            list.get(j).add(i);
    }

    public void printGraph() {
        for (int i = 0; i < this.vertices; i++) {
            System.out.print(i);
            for (int j = 0; j < this.list.get(i).size(); j++){
                System.out.print(" -> ");
                System.out.print(this.list.get(i).get(j));
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        GraphLinkedList graph = new GraphLinkedList(5, GraphType.UNDIRECTED);
        graph.addEdge(0, 1);
        graph.addEdge(1, 2);
        graph.addEdge(1, 4);
        graph.printGraph();
    }
}
